src/states/adult/mod.rs

```diff
@@ -527,7 +497,15 @@ impl Base for Adult {
 
         for msg in mem::replace(&mut self.routing_msg_backlog, Default::default()) {
             if let Transition::Stay = &transition {
-                match self.handle_backlogged_filtered_signed_message(msg) {
+                let msg = match HopMessageWithBytes::new(msg) {
+                    Ok(msg) => msg,
+                    Err(err) => {
+                        error!("{} - Failed to make message {:?}", self, err);
+                        continue;
+                    }
+                };
+
+                match self.handle_filtered_signed_message(msg) {
```

Comment:
This will forward messages whose trust may be outdated.
I think the idea was that we did not need to do that, and probably don't want to do that.

Is it to handle the JoiningPeer needing to forward the message, it already has a set of elders so it could forward message there like Adult do, instead of on transition, that would probably be better.

But probably ideally we would be to ensure that only Adults are transistion to Elders. i.e something that would stop a node being promoted if we have not detected that it transitioned to Adult already?

Reply:
> Is it to handle the JoiningPeer needing to forward the message, it already has a set of elders so it could forward message there like Adult do, instead of on transition, that would probably be better.

This sounds like decent idea that can make things simpler. I'm going to give it a try.

> But probably ideally we would be to ensure that only Adults are transistion to Elders. i.e something that would stop a node being promoted if we have not detected that it transitioned to Adult already?

I'm not sure I understand this. We do transition to `Elders` only from `Adults`. Or am I missing something?

Reply:
> I'm not sure I understand this. We do transition to Elders only from Adults. Or am I missing something?

Reply:
> I'm not sure I understand this. We do transition to Elders only from Adults. Or am I missing something?

Yes, the nodes go through Joining -> Adult -> Elder, however, a section may promote to Elder a node that still think of itself as Joining. Is that not the source of the bug?

Reply:
Ah yes, got you now. Yes, it is the source of the bug, but I don't think it's buggy by itself because it's just the node being a bit laggy.

Reply:
Indeed, what could be considered buggy is that we decide a node is an Elder before checking they can do it (i.e are already Adults, primed and ready to transition when notified).

Reply:
But that would involve the adult to check some kind of acknowledgement to the elders before they proceed with the promotion. Do we really want that?

Reply:
Yes, indeed, we already have Adult sending Knowledge information/Poke so we could use that.
Whether we want to or not is open question, simplify some things complicate others.

Reply:
My thought though, is that it probably is something we want because it ensures we know what kind of node receive a message, thus avoiding having to push more and more things into the JoiningPeer type.

Reply:
I think this is a potential can of worms. We can use the `members_knowledge` container to know which nodes have caught up and only promote those that have. But that container is not shared state which means we can't base voting on it. So we would have to make it shared state with all the complications that it entails.

Reply:
Yes, we could not use the `members_knowledge` container, we could vote for node as Adults when we get their message though, and update the MemberInfo for it when it reach consensus (i.e like we do for Online).

But yes it could have further problem, still incline to think it will be a better solution in the long run, but does not have to be in that PR if we think the alternative is simpler to live with for now.

Reply:
OK. If we decide to do this, then I'd strongly prefer to do it in a separate PR. Anyway, back to the point:
> Is it to handle the JoiningPeer needing to forward the message, it already has a set of elders so it could forward message there like Adult do, instead of on transition, that would probably be better.

I tried this but it doesn't work. The problem is that when we are `JoiningPeer`, we don't yet have enough information to decide whether to handle a message or relay it. We can only do that when we become `Adult`. So the only sensible thing to do as `JoiningPeer` is to backlog it.At least until we come up with different approach than backlogging everything.

Reply:
> This will forward messages whose trust may be outdated.

So to handle this correctly, I think we need different logic for handling the `JoiningPeer -> Adult` transition from the `Elder` -> `Adult` one.

Reply:
> OK. If we decide to do this, then I'd strongly prefer to do it in a separate PR. Anyway, back to the point:

It could be a separate PR coming first if we cannot address the issues with having JoiningNode getting messages too early (but lets see what we can do without needing it).

> The problem is that when we are JoiningPeer, we don't yet have enough information to decide whether to handle a message or relay it.

What do you mean, I thought we would be relaying it always since we are a JoiningPeer, and also push it in our backlog for processing?

Reply:
Ah, so backlog AND relay. OK, let's try that.

Reply:
Yes, this is what Adults do too, i.e they backlog for processing (but not re-relay), and relay message immediately. It also avoid that message being delayed.

Reply:
OK, that makes sense. However is still doesn't work. At least now if `Adult::handle_backlogged_filtered_signed_message` is implemented the way it currently is. Problem is we still need to backlog a message that is meant for us as elder, otherwise when we get promoted the message gets lost. Even if someone would then send the message to us again, it would get filtered out because we've already seen it (This might point to a deeper problem in that if we don't handle the message ourselves, we should probably not incoming-filter it). 

Also:
> This will forward messages whose trust may be outdated.

I don't think this is a problem, because we clear the backlog on genesis update. 

So I'm now thinking reverting to the original approach - that is - call `handle_filtered_signed_message` in both cases (new and backlogged). Note this is different from what we do as `Elder` but it makes sense because as `Elder` we always process the message, but as `Adult` we don't.

Reply:
Also, I don't think it makes much sense to relay messages as `JoiningPeer` because we don't expect to get stuck in that state for long, unlike `Adult`.

Reply:
> OK, that makes sense. However is still doesn't work. At least now if Adult::handle_backlogged_filtered_signed_message is implemented the way it currently is.

Is it that we drop the message when we become Elder in `Adult::handle_backlogged_filtered_signed_message`, are we not at the `in_location`? why?

Reply:
> Also, I don't think it makes much sense to relay messages as JoiningPeer because we don't expect to get stuck in that state for long, unlike Adult.

I'm confused, I thought that was the bug, that JoiningPeer received message that Elder should have received, is it not?

Reply:
Let me summarize this:

The problem we have is that a `JoiningPeer` receives a message meant for `Elder` because it's lagging. The simplest solution to deal with this and the one currently implemented in this PR is this:
1. Backlog the message as `JoiningPeer`
2. When transitioning into `Adult`, pull the messages from the backlog, process those we can and re-backlog the others
3. When transitioning into `Elder`, pull the messages from the backlog and process them all

The main difference from the current fleming branch is the re-backlogging of message on transition to `Adult` which we don't do in fleming. This is because in fleming we have two pathways for handling messages: one for received messages and one for messages from backlog. They differ mostly in what we do with unhandled messages - in the first one we backlog them, in the second one we ignore them. 

Notes about some of your suggestions:
- Relaying messages also as `JoiningPeer` - we can do this but it's not going to buy us much at the cost of complicating the `JoiningPeer` state. This is because we don't stay in that state for long and we do relay the messages when transitioning to `Adult`. Also this wouldn't be enough to fix the primary failure anyway, so I decided to not implement it in this PR.
- Not re-backlogging messages when transitioning to `Adult` - already explained that this is the source of the failure. The concern about message becoming stale this way is unfounded, because we purge the backlog on every `GenesisUpdate`.